controllers/eventhandlers/targetgrouppolicy.go

```diff
@@ -0,0 +1,149 @@
+package eventhandlers
+
+import (
+	"context"
+	"github.com/aws/aws-application-networking-k8s/pkg/apis/applicationnetworking/v1alpha1"
+	"github.com/aws/aws-application-networking-k8s/pkg/model/core"
+	"github.com/aws/aws-application-networking-k8s/pkg/utils/gwlog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	gateway_api_v1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gateway_api "sigs.k8s.io/gateway-api/apis/v1beta1"
+	mcs_api "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+)
+
+type targetGroupPolicyEventHandler struct {
+	log    gwlog.Logger
+	client client.Client
+}
+
+func NewTargetGroupPolicyEventHandler(log gwlog.Logger, client client.Client) *targetGroupPolicyEventHandler {
+	return &targetGroupPolicyEventHandler{log: log, client: client}
+}
+
+func (h *targetGroupPolicyEventHandler) getTargetRef(obj client.Object) *corev1.Service {
+	tgp := obj.(*v1alpha1.TargetGroupPolicy)
+	policyName := tgp.Namespace + "/" + tgp.Name
+
+	targetRef := tgp.Spec.TargetRef
+	if targetRef.Group != "" || targetRef.Kind != "Service" {
+		h.log.Warnw("Detected non-Service TargetGroupPolicy attachment, skipping",
+			"policyName", policyName, "targetRef", targetRef)
+		return nil
+	}
+	namespace := tgp.Namespace
+	if targetRef.Namespace != nil && namespace != string(*targetRef.Namespace) {
+		h.log.Warnw("Detected cross namespace TargetGroupPolicy attachment, skipping",
+			"policyName", policyName, "targetRef", targetRef)
+		return nil
+	}
+
+	svcName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      string(targetRef.Name),
+	}
+	svc := &corev1.Service{}
+	if err := h.client.Get(context.TODO(), svcName, svc); err != nil {
+		return nil
```

Comment:
should return *Service and error? it seems nil can be a good "nothing to return" and bad "error"

Reply:
* The function's type is restricted to k8s-provided `handler.MapFunc` so that we cannot handle errors without panics
* I think it is better to gracefully handle the error and just skip the event. Instead we can put error-level logs here.

Reply:
> The function's type is restricted to k8s-provided handler.MapFunc so that we cannot handle errors without panics

I don't think ```getTargetRef``` is subject to this restriction? I see how the ```Map*``` functions would be.

Reply:
I guess the result is the same, we return nil from our mapping function

Reply:
Yes, but we intentionally swallow error here so I prefer not returning it (just to swallow later). we have more information inside and we can see targetRef related logs put in the same place